README
mysql module for Qore
by David Nichols

Requires qore 0.8.6+ and mysql 4.1+ headers and libraries for the full driver feature set

With MySQL 4.1+ you can get transaction support and the module will use the more efficient prepared statement interface.

The driver supports the following features (depending on the MySQL client library version):
* is thread-safe
* stored procedure execution with binding and retrieving values (5 and newer)
* transaction management is supported (4.1 and newer)
* transparent character encoding conversion is supported if necessary
* the "optimal-numbers" option is set by default as of module version 2.2
* supports a server timezone setting for correctly handling date/time value
  when communicating with a DB server in another time zone
* the prepared statement API is support (SQLStatement class)

The driver's name is "mysql" and the Datasource type constant is SQL::DSMySQL
ex: my Datasource $db(DSMySQL);
ex: my Datasource $db("mysql:user/pass@db%hostname:3306");

The mysql module is stable and has been extensively tested.

Documentation: docs/mysql-module-doc.html

test/example script: test/db-test.q

BUILD AND INSTALL
-----------------

If your mysql installation is in a non-standard location, set the MYSQL_DIR environment variable to the location of the installation before running configure.
Note that you have to use g++ 4.0.* on Darwin to link with newer versions of the MySQL libraries

configure
make
sudo make install

The configure script will find out where your qore module directory is found and set this for the install directory.

BUILD FROM SVN
--------------

reconf.sh
configure
make
sudo make install


ChangeLog
2012-11-26 07:13  david_nichols

	* ChangeLog, README: updates

2012-11-18 16:39  david_nichols

	* docs/mainpage.dox.tmpl: updated docs for new version

2012-11-18 16:34  david_nichols

	* RELEASE-NOTES, configure.ac, qore-mysql-module.spec,
	  src/mysql.cpp: updated module version to 2.0.1 - fixed a bug in
	  returning numeric values where numbers with a decimal component
	  were being returned as integers when the default option
	  "optimal-numbers" was enabled

2012-10-28 18:38  david_nichols

	* ChangeLog: updated changelog

2012-10-28 18:38  david_nichols

	* README, RELEASE-NOTES, qore-mysql-module.spec: doc/spec updates
	  for 2.0 release

2012-10-20 15:58  david_nichols

	* configure.ac: windows build fixes

2012-09-29 07:14  david_nichols

	* RELEASE-NOTES, docs/mainpage.dox.tmpl, src/mysql.cpp,
	  src/qore-mysql.h: added support for DBI_CAP_SERVER_TIME_ZONE and
	  the "timezone" option; return all date/time values in the
	  server's time zone, bind all date/times in the server's time zone

2012-09-28 16:28  david_nichols

	* src/mysql.cpp: fixed a memory error related the new prepared
	  statement support

2012-09-28 13:24  david_nichols

	* docs/mainpage.dox.tmpl: updated docs

2012-09-28 13:20  david_nichols

	* qore-mysql-module.spec: updated spec file for doc package and new
	  doxygen-based html docs

2012-09-28 13:17  david_nichols

	* Makefile.am, docs/doxyfile.tmpl, docs/footer_template.html,
	  docs/mainpage.dox.tmpl, docs/mysql-module-doc.html, docs/old,
	  docs/old/mysql-module-doc.html: created doxygen docs

2012-09-28 11:24  david_nichols

	* test/sql-stmt.q: added prepared statement test script

2012-09-28 11:16  david_nichols

	* RELEASE-NOTES, src/mysql.cpp, src/qore-mysql.h: implemented
	  support for the selectRow DBI method
	  implemented supoprt for the prepared statement API
	  fixed memory errors with selecting binary data types

2012-09-26 15:37  david_nichols

	* RELEASE-NOTES, configure.ac, qore-mysql-module.spec,
	  src/Makefile.am, src/mysql.cc, src/mysql.cpp, src/qore-mysql.h: *
	  updated module version to 2.0
	  * added numeric and dbi option support when compiled with qore
	  0.8.6+

2012-06-25 08:51  david_nichols

	* configure.ac: updated configure.ac logic to deal with /lib ->
	  /usr/lib and /lib64 -> /usr/lib64 as with Fedora 17 - the
	  previous logic was too clever for its own good

2012-05-22 16:04  david_nichols

	* configure.ac: an initial attempt to get the mysql module to build
	  for windows with mxe (mingw cross-compiler)

2012-01-16 13:07  david_nichols

	* configure.ac: updated configure scripts to only set LIBSUFFIX=64
	  on 64-bit platfiorms if /usr/lib64 exists

2011-09-12 16:26  david_nichols

	* configure.ac: added -mt to compile and link lines when building
	  on solaris with CC

2010-11-09 10:06  david_nichols

	* src/qore-mysql.h: fixed reconnect logic

2010-11-08 16:47  david_nichols

	* src/qore-mysql.h: fixed mysql and sybase auto-reconnect logic to
	  use the new Datasource::activeTransaction() API method

2010-06-21 09:12  david_nichols

	* qore-mysql-module.spec: updated spec files to generically
	  determine *suse versions for distribution tagging in rpm file
	  names

2010-06-20 06:00  david_nichols

	* ChangeLog: updated changelog with fixed svn2cl

2010-06-20 05:49  david_nichols

	* README-BINARIES: added readme

2010-06-19 18:56  david_nichols

	* qore-mysql-module.spec: minor rpm updates

2010-06-19 18:53  david_nichols

	* ChangeLog: updated changelog

2010-06-15 18:49  david_nichols

	* docs/mysql-module-doc.html: updated tikiwiki and added updated
	  color scheme

2010-06-10 14:23  david_nichols

	* RELEASE-NOTES: updated release notes for new release

2010-05-12 11:49  david_nichols

	* configure.ac: solaris and sun studio compile and warning
	  fixes/enhancements

2010-05-11 14:44  david_nichols

	* reconf.sh: compile fixes for solaris

2010-04-28 10:15  david_nichols

	* test/db-test.q: updates for new qore API and to address new
	  warnings in test scripts

2010-04-21 11:14  david_nichols

	* src/mysql.cc, src/qore-mysql.h: updated to use more efficient
	  APIs in qore 0.8+ if available when binding date values

2010-04-16 09:32  subik

	* RELEASE-NOTES, configure.ac, qore-mysql-module.spec,
	  src/mysql.cc: sql drivers versions bump to spot execRaw. execRaw
	  feature promoted to all drivers' feature lists. execRaw constant
	  naming unified.

2010-02-22 11:46  subik

	* src/mysql.cc: execRaw for mysql

2009-12-09 12:58  david_nichols

	* configure.ac: fixed all modules to use decimal comparisons with
	  the qore version string instead of octal so that they will still
	  compile with qore v0.8.0 and up...

2009-11-22 06:03  david_nichols

	* configure.ac: updated configure to assume 64-bit build if
	  host_cpu = "amd64"

2009-10-30 08:22  david_nichols

	* configure.ac: updated all module configure scripts to use gnutar
	  on darwin to avoid picking up OSX resource files in source
	  releases

2009-10-03 09:56  david_nichols

	* RELEASE-NOTES: cosmetic change

2009-09-22 12:29  david_nichols

	* RELEASE-NOTES, configure.ac, qore-mysql-module.spec,
	  src/mysql.cc: fixed setting transaction isolation level; was
	  broken when mysql 5.1 fixed a
	  bug; updated release to 1.0.7

2009-09-22 12:18  david_nichols

	* src/mysql.cc: cosmetic change

2009-09-21 11:07  david_nichols

	* configure.ac: updated module configure scripts to select 64-bit
	  target mode by default on snow leopard i386

2009-09-10 06:49  david_nichols

	* configure.ac: updated configure.ac to properly compile in 32-bit
	  mode on OSX snow leopard (10.6)

2009-07-22 16:00  david_nichols

	* docs/mysql-module-doc.html: updated docs to reflect the
	  possibilty to give a port number

2009-07-22 15:55  david_nichols

	* ChangeLog: updated changelog

2009-07-20 13:16  david_nichols

	* qore-mysql-module.spec: added opensuse 11.2 detection to spec
	  file, fixed rhel detection in spec files, updated configure.ac to

src/qore-mysql-module.h
/*
  mysql-module.h

  MYSQL integration to QORE

  Copyright 2003 - 2009 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef _QORE_MYSQL_MODULE_H

#define _QORE_MYSQL_MODULE_H

class QoreStringNode *qore_mysql_module_init();
void qore_mysql_module_ns_init(class QoreNamespace *rns, class QoreNamespace *qns);
void qore_mysql_module_delete();

#endif


src/qore-mysql.h
/* -*- mode: c++; indent-tabs-mode: nil -*- */
/*
  qore-mysql.h

  Qore Programming Language

  Copyright (C) 2003 - 2012 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef _QORE_MYSQL_H

#define _QORE_MYSQL_H

#include "../config.h"

#include <qore/Qore.h>

#include <mysql.h>

#ifndef MYSQL_PORT
#define MYSQL_PORT 3306
#endif

#ifdef HAVE_MYSQL_STMT
class QoreMysqlConnection;

static inline void get_lower_case_name(QoreString* str, const QoreEncoding* enc, const char *name) {
   str->set(name, enc);
   str->tolwr();
}

class MyResult {
private:
   MYSQL_FIELD *field;
   int num_fields;
   MYSQL_BIND *bindbuf;
   struct bindInfo {
      my_bool mnull;
      long unsigned int mlen;
   } *bi;
   QoreMysqlConnection* conn;
   const QoreEncoding* enc;

public:
   DLLLOCAL MyResult(QoreMysqlConnection* c, const QoreEncoding* n_enc)
      : field(0), num_fields(0), bindbuf(0), bi(0), conn(c), enc(n_enc) {
   }

   DLLLOCAL ~MyResult() {
      reset();
   }

   DLLLOCAL void reset() {
      if (bindbuf) {
	 // delete buffer
	 for (int i = 0; i < num_fields; i++) {
            switch (bindbuf[i].buffer_type) {
               case MYSQL_TYPE_DOUBLE:
               case MYSQL_TYPE_LONGLONG:
               case MYSQL_TYPE_STRING:
               case MYSQL_TYPE_BLOB:
                  if (bindbuf[i].buffer)
                     free(bindbuf[i].buffer);
                  break;

               case MYSQL_TYPE_DATETIME:
                  delete (MYSQL_TIME*)bindbuf[i].buffer;
                  break;

               // to avoid warnings
               default:
                  break;
            }
         }
	 delete [] bindbuf;
         bindbuf = 0;
      }
      if (bi) {
	 delete [] bi;
         bi = 0;
      }
   }

   DLLLOCAL void set(MYSQL_STMT* stmt) {
      reset();
      MYSQL_RES* res = mysql_stmt_result_metadata(stmt);
      if (res) {
         field = mysql_fetch_fields(res);
         num_fields = mysql_num_fields(res);
         mysql_free_result(res);
      }
   }

   DLLLOCAL operator bool() const {
      return (bool)num_fields;
   }

   DLLLOCAL void bind(MYSQL_STMT *stmt);
   DLLLOCAL AbstractQoreNode* getBoundColumnValue(int i, bool destructive = false);

   DLLLOCAL char *getFieldName(int i) {
      return field[i].name;
   }

   DLLLOCAL int getNumFields() {
      return num_fields;
   }

   DLLLOCAL QoreHashNode* getSingleRow(ExceptionSink* xsink) {
      QoreHashNode* h = new QoreHashNode;

      for (int i = 0; i < num_fields; i++) {
         QoreString tstr(field[i].name, enc);
         tstr.tolwr();
         h->setKeyValue(&tstr, getBoundColumnValue(i), xsink);
      }
      return h;
   }

   // returns a hash of empty lists keyed by column name
   DLLLOCAL QoreHashNode* setupColumns() {
      QoreHashNode* h = new QoreHashNode;
      QoreString tstr;
      for (int i = 0; i < num_fields; i++) {
	 get_lower_case_name(&tstr, enc, field[i].name);
	 h->setKeyValue(&tstr, new QoreListNode, 0);
      }

      return h;
   }
};

// FIXME: do not assume byte widths
union my_val {
   MYSQL_TIME time;
   int i4;
   int64 i8;
   double f8;
   void *ptr;

   DLLLOCAL void assign(const QoreMysqlConnection& conn, const DateTime &d);
};

class QoreMysqlBindNode {
protected:
   DLLLOCAL ~QoreMysqlBindNode() {
      assert(!data.value);
      assert(!data.tstr);
   }

public:
   int bindtype;
   unsigned long len;

   struct {
      AbstractQoreNode *value;   // value to be bound
      QoreStringNode *tstr;   // temporary string to be deleted
   } data;

   union my_val vbuf;
   QoreMysqlBindNode *next;

   // for value nodes
   DLLLOCAL QoreMysqlBindNode(const AbstractQoreNode *v) {
      bindtype = BN_VALUE;
      data.value = v ? v->refSelf() : 0;
      data.tstr = 0;
      next = 0;
   }

   DLLLOCAL void del(ExceptionSink* xsink) {
      reset(xsink);
      delete this;
   }

   DLLLOCAL int reset(ExceptionSink* xsink) {
      if (data.tstr) {
	 data.tstr->deref();
         data.tstr = 0;
      }
      if (data.value) {
         data.value->deref(xsink);
         data.value = 0;
      }
      return *xsink ? -1 : 0;
   }

   DLLLOCAL int rebind(const AbstractQoreNode* v, ExceptionSink* xsink) {
      if (reset(xsink))
         return -1;
      data.value = v ? v->refSelf() : 0;
      return 0;
   }
     
   DLLLOCAL int bindValue(const QoreMysqlConnection& conn, MYSQL_BIND *buf, ExceptionSink* xsink);
};

static MYSQL *qore_mysql_init(Datasource *ds, ExceptionSink* xsink);

static inline bool wasInTransaction(Datasource *ds) {
#ifdef _QORE_HAS_DATASOURCE_ACTIVETRANSACTION
   return ds->activeTransaction();
#else
   return ds->isInTransaction();
#endif
}

#define OPT_NUM_OPTIMAL 0  // return numeric as int64 if it fits or "number" if not
#define OPT_NUM_STRING  1  // always return numeric types as strings
#define OPT_NUM_NUMERIC 2  // always return numeric types as "number"

#ifdef _QORE_HAS_DBI_OPTIONS
// return optimal numeric values if options are supported
#define OPT_NUM_DEFAULT OPT_NUM_OPTIMAL
#else
// return numeric values as strings if options are not supported -- for backwards-compatibility
#define OPT_NUM_DEFAULT OPT_NUM_STRING
#endif

class QoreMysqlConnection {
public:
   MYSQL* db;
   Datasource& ds;
#ifdef _QORE_HAS_FIND_CREATE_TIMEZONE
   const AbstractQoreZoneInfo* server_tz;
#endif
   int numeric_support;

   DLLLOCAL QoreMysqlConnection(MYSQL* d, Datasource& n_ds) 
      : db(d), ds(n_ds), 
#ifdef _QORE_HAS_FIND_CREATE_TIMEZONE
        server_tz(currentTZ()),
#endif
        numeric_support(OPT_NUM_DEFAULT) {
   }

   DLLLOCAL ~QoreMysqlConnection() {
      mysql_close(db);
   }

   DLLLOCAL int reconnect(Datasource *ds, MYSQL_STMT *&stmt, const QoreString& str, ExceptionSink* xsink) {	 
      // throw an exception if a transaction is in progress
      if (wasInTransaction(ds))
	 xsink->raiseException("DBI:MYSQL:CONNECTION-ERROR", "connection to MySQL database server lost while in a transaction; transaction has been lost");

      MYSQL *new_db = qore_mysql_init(ds, xsink);
      if (!new_db) {
         ds->connectionAborted();
	 return -1;
      }

      printd(5, "mysql datasource %08p reconnected after timeout\n", ds);
      mysql_close(db);
      db = new_db;

      if (wasInTransaction(ds))
         return -1;

      // reinitialize statement
      mysql_stmt_close(stmt);
      stmt = stmt_init(xsink);
      if (!stmt)
	 return -1;
	 
      // prepare the statement for execution (again)
      if (mysql_stmt_prepare(stmt, str.getBuffer(), str.strlen()))
	 return -1;
	 
      return 0;
   }

   DLLLOCAL int commit() {
      return mysql_commit(db);
   }

   DLLLOCAL int rollback() {
      return mysql_rollback(db);
   }

   DLLLOCAL const char *error() {
      return mysql_error(db);
   }

   DLLLOCAL int q_errno() {
      return mysql_errno(db);
   }

   DLLLOCAL MYSQL_STMT *stmt_init(ExceptionSink* xsink) {
      MYSQL_STMT *stmt = mysql_stmt_init(db);
      if (!stmt)
	 xsink->raiseException("DBI:MYSQL:ERROR", "error creating MySQL statement handle: out of memory");
      return stmt;
   }

   DLLLOCAL unsigned long getServerVersion() {
      return mysql_get_server_version(db);
   }

   DLLLOCAL int setOption(const char* opt, const AbstractQoreNode* val, ExceptionSink* xsink) {
      if (!strcasecmp(opt, DBI_OPT_NUMBER_OPT)) {
         numeric_support = OPT_NUM_OPTIMAL;
         return 0;
      }
      if (!strcasecmp(opt, DBI_OPT_NUMBER_STRING)) {
         numeric_support = OPT_NUM_STRING;
         return 0;
      }
      if (!strcasecmp(opt, DBI_OPT_NUMBER_NUMERIC)) {
         numeric_support = OPT_NUM_NUMERIC;
         return 0;
      }
#ifdef _QORE_HAS_FIND_CREATE_TIMEZONE
      assert(!strcasecmp(opt, DBI_OPT_TIMEZONE));
      assert(get_node_type(val) == NT_STRING);
      const QoreStringNode* str = reinterpret_cast<const QoreStringNode*>(val);
      const AbstractQoreZoneInfo* tz = find_create_timezone(str->getBuffer(), xsink);
      if (*xsink)
         return -1;
      server_tz = tz;
#else
      assert(false);
#endif
      return 0;
   }

   DLLLOCAL AbstractQoreNode* getOption(const char* opt) {
      if (!strcasecmp(opt, DBI_OPT_NUMBER_OPT))
         return get_bool_node(numeric_support == OPT_NUM_OPTIMAL);

      if (!strcasecmp(opt, DBI_OPT_NUMBER_STRING))
         return get_bool_node(numeric_support == OPT_NUM_STRING);

      if (!strcasecmp(opt, DBI_OPT_NUMBER_NUMERIC))
         return get_bool_node(numeric_support == OPT_NUM_NUMERIC);

#ifdef _QORE_HAS_FIND_CREATE_TIMEZONE
      assert(!strcasecmp(opt, DBI_OPT_TIMEZONE));
      return new QoreStringNode(tz_get_region_name(server_tz));
#else
      assert(false);
#endif
      return 0;
   }

   DLLLOCAL int getNumeric() const { 
      return numeric_support; 
   }

#ifdef _QORE_HAS_TIME_ZONES
   DLLLOCAL const AbstractQoreZoneInfo* getTZ() const {
#ifdef _QORE_HAS_FIND_CREATE_TIMEZONE
      return server_tz;
#else
      return currentTZ();
#endif
   }
#endif
};

class QoreMysqlBindGroup {
protected:
   QoreMysqlBindNode *head, *tail;
   QoreString *str;
   MYSQL_STMT *stmt;
   bool hasOutput;
   MYSQL_BIND *bind;
   Datasource *ds;
   QoreMysqlConnection* mydata;
   int len;
   cstr_vector_t phl;
   MyResult myres;

   // returns -1 = error, 0 = OK, 1 = server doesn't support prepared statements
   DLLLOCAL int prepare(bool unsupported_ok, ExceptionSink* xsink);

   // returns -1 = error, 0 = OK
   DLLLOCAL int rebindArgs(const QoreListNode* args, ExceptionSink* xsink);
   DLLLOCAL int bindArgs(ExceptionSink* xsink);

   // returns -1 = error, 0 = OK
   DLLLOCAL inline int parse(const QoreListNode *args, ExceptionSink* xsink);
   DLLLOCAL inline void add(class QoreMysqlBindNode *c) {
      len++;
      if (!tail)
	 head = c;
      else
	 tail->next = c;
      tail = c;
   }

   DLLLOCAL int execIntern(ExceptionSink* xsink);
   
   DLLLOCAL int getDataRows(QoreListNode& l, ExceptionSink* xsink, int max = -1);
   DLLLOCAL int getDataColumns(QoreHashNode& h, ExceptionSink* xsink, int max = -1);

   DLLLOCAL ~QoreMysqlBindGroup() {
      assert(!head);
      assert(!str);
      assert(!stmt);
      assert(!bind);
   }

public:
   DLLLOCAL QoreMysqlBindGroup(Datasource *ods) 
   : head(0), tail(0), str(0), stmt(0), hasOutput(false), bind(0), 
     ds(ods), mydata((QoreMysqlConnection *)ds->getPrivateData()), len(0),
     myres(mydata, ds->getQoreEncoding()) {
   }

   DLLLOCAL void del(ExceptionSink* xsink) {
      reset(xsink);
      delete this;
   }

   DLLLOCAL void reset(ExceptionSink* xsink);

   // returns 0=OK, -1=error and exception raised, 1=statement cannot be prepared
   DLLLOCAL int prepareAndBind(const QoreString *ostr, const QoreListNode *args, ExceptionSink* xsink);

   DLLLOCAL void add(const AbstractQoreNode *v) {
      add(new QoreMysqlBindNode(v));
      printd(5, "QoreMysqlBindGroup::add() value=%08p\n", v);
   }

   DLLLOCAL void add(char *name) {
      phl.push_back(name);
      printd(5, "QoreMysqlBindGroup::add() placeholder '%s' %d %s\n", name);
      hasOutput = true;
   }

   DLLLOCAL int affectedRows() const {
      assert(stmt);
      return mysql_stmt_affected_rows(stmt);
   }

   // also can be used like "select"
   DLLLOCAL AbstractQoreNode* exec(ExceptionSink* xsink);
   DLLLOCAL AbstractQoreNode* selectRows(ExceptionSink* xsink);
   DLLLOCAL QoreHashNode* selectRow(ExceptionSink* xsink);

   DLLLOCAL QoreHashNode* getOutputHash(ExceptionSink* xsink);
};

class QoreMysqlBindGroupHelper : public QoreMysqlBindGroup {
protected:
   ExceptionSink* xsink;

public:
   DLLLOCAL QoreMysqlBindGroupHelper(Datasource* ds, ExceptionSink* xs) : QoreMysqlBindGroup(ds), xsink(xs) {
   }

   DLLLOCAL ~QoreMysqlBindGroupHelper() {
      reset(xsink);
   }
};

#ifdef _QORE_HAS_PREPARED_STATMENT_API
class QoreMysqlPreparedStatement : public QoreMysqlBindGroup {
public:
   DLLLOCAL QoreMysqlPreparedStatement(Datasource* ds) : QoreMysqlBindGroup(ds) {
   }

   DLLLOCAL ~QoreMysqlPreparedStatement() {
   }

   // returns 0 for OK, -1 for error
   DLLLOCAL int prepare(const QoreString& sql, const QoreListNode* args, bool parse, ExceptionSink* xsink);
   DLLLOCAL int bind(const QoreListNode& l, ExceptionSink* xsink);
   DLLLOCAL int define(ExceptionSink* xsink);
   DLLLOCAL int exec(ExceptionSink* xsink);
   DLLLOCAL QoreHashNode* fetchRow(ExceptionSink* xsink);
   DLLLOCAL QoreListNode* fetchRows(int rows, ExceptionSink* xsink);
   DLLLOCAL QoreHashNode* fetchColumns(int rows, ExceptionSink* xsink);
   DLLLOCAL bool next();
};
#endif // _QORE_HAS_PREPARED_STATMENT_API

#endif // HAVE_MYSQL_STMT

#endif // _QORE_MYSQL_H
